### ISSUE TITLE
uwsgi tries to load ./python3_plugin.so, ignoring plugins-dir option

### ISSUE THREAD

**Issue by PDiracDelta** [unbit/uwsgi#1992]
I run uwsgi 2.0.17 on Ubuntu server 16.04 with my config script
`/usr/local/bin/uwsgi --ini uwsgi_config_new.ini`
but it fails, stating

> [uWSGI] getting INI configuration from /srv/QCQuan_new/uwsgi_config_new.ini
> open("./python3_plugin.so"): No such file or directory [core/utils.c line 3713]
> !!! UNABLE to load uWSGI plugin: ./python3_plugin.so: cannot open shared object file: No such file or directory !!!

However, the plugin does exist:
`ls /usr/lib/uwsgi/plugins/python3_plugin.so `

> /usr/lib/uwsgi/plugins/python3_plugin.so

And I am using the correct(?) `plugins-dir` option:
`cat /srv/QCQuan_new/uwsgi_config_new.ini | grep plugin`

> plugins-dir = /usr/lib/uwsgi/plugins/
> plugins = python3

Still, uwsgi seems to want to load the `python3_plugin.so` from the current directory `.` instead of the specified directory. And why doesn't it default to `/usr/lib/uwsgi/plugins` anyway?.

Update: it doesn't help to place a link to the plugin file in the dir from which you start uwsgi, nor in the dir specified by the `chdir` option in the config.

Update 2020-05-08: It keeps wanting to start python2.7 even though I clearly ask it to use `python36_plugin.so` (yeah, meanwhile, I upgraded from 3 to 3.6).

**Comment by dgilge** [comment]
Did you try `plugins-dir = /usr/lib/uwsgi`?

**Comment by PDiracDelta** [comment]
Yes, that doesn't work either, but I now see it gives slightly different behaviour: when I try `plugins-dir = /usr/lib/uwsgi` it seems to try both that dir as well as `.`:

> [uWSGI] getting INI configuration from /srv/QCQuan_new/uwsgi_config_new.ini
> uwsgi[12346]: open("/usr/lib/uwsgi/python3_plugin.so"): No such file or directory [core/utils.c line 3713]
> uwsgi[12346]: open("./python3_plugin.so"): No such file or directory [core/utils.c line 3713]
> uwsgi[12346]: !!! UNABLE to load uWSGI plugin: ./python3_plugin.so: cannot open shared object file: No such file or directory !!!

From `open("/usr/lib/uwsgi/python3_plugin.so")` I understand that `plugins-dir` should really include the directory `plugins` in the path. However if I specify the path as such, it seems to refuse to check the path (or at least it doesn't write a line to the log) as you can see from my first post.

**Comment by natefoo** [comment]
I suspect this is a red herring - that with your original setting it was in fact finding the plugin in `/usr/lib/uwsgi/plugins` but failing for other reasons. In my case it was because I had zerg mode enabled but the zerg pool was down.

**Comment by SushantDaga** [comment]
Hi, I was getting the same error.
The following works for me:
```
[uwsgi]
plugins-dir = /usr/lib/uwsgi/
plugins = /usr/lib/uwsgi/plugins/python36_plugin.so
```
I guess removing plugins-dir might work too. Make sure there is no extra ```plugins = python3``` statements in ini file.
Does this solve the issue?

**Comment by PDiracDelta** [comment]
Unfortunately that doesn't work for me.
It keeps wanting to start python2.7 even though I clearly ask it to use `python36_plugin.so` (yeah, meanwhile, I upgraded from 3 to 3.6). (from the uwsgi log):

> Python version: 2.7.12 (default, Apr 15 2020, 17:07:12)  [GCC 5.4.0 20160609]
> Python main interpreter initialized at 0x2258ae0

... after which it can't find my app, of course, because it is only installed in python3(.6)

**Comment by xrmx** [comment]
@PDiracDelta you have uwsgi in /usr/local and plugins in /usr, it doesn't look right, looks like you two uwsgi installations and mixing them

**Comment by PDiracDelta** [comment]
> @PDiracDelta you have uwsgi in /usr/local and plugins in /usr, it doesn't look right, looks like you two uwsgi installations and mixing them

Well spotted! I had no idea. I investigated some more and I discovered that (probably?) the binary I was using comes pre-compiled with python(2) as the plugin it should load.
I purged all my installs and followed [this guide](https://serverfault.com/a/916780/383377) to make it work properly. It builds uwsgi with 'nolang' specified and then builds the python3 plugin afterwards. Then, I can use it as is described in the docs by using `plugins-dir=` and `plugin=`
